source/jormungandr/jormungandr/street_network/geovelo.py

```diff
@@ -0,0 +1,190 @@
+
+
+class Geovelo(AbstractStreetNetworkService):
+    sn_system_id = 'geovelo'
+
+    def __init__(self, instance, service_url, timeout=10, api_key=None, **kwargs):
+        self.instance = instance
+        if not is_url(service_url):
+            raise ValueError('service_url is invalid, you give {}'.format(service_url))
+        self.service_url = service_url
+        self.api_key = api_key
+        self.timeout = timeout
+        self.breaker = pybreaker.CircuitBreaker(fail_max=app.config['CIRCUIT_BREAKER_MAX_GEOVELO_FAIL'],
+                                                reset_timeout=app.config['CIRCUIT_BREAKER_GEOVELO_TIMEOUT_S'])
+
+    def _pt_object_summary(self, pt_object):
+        coord = get_pt_object_coord(pt_object)
+        return [coord.lat, coord.lon, getattr(pt_object, 'uri', None)]
+
+    def _make_data(self, origins, destinations):
+        data = {}
+        data['starts'] = []
+        for o in origins:
+            data['starts'].append(self._pt_object_summary(o))
+        data['ends'] = []
+        for d in destinations:
+            data['ends'].append(self._pt_object_summary(d))
+        return data
+
+    def _call_geovelo(self, url, method=requests.post, data=None):
+        logging.getLogger(__name__).debug('Geovelo routing service , call url : {}'.format(url))
+        try:
+            return self.breaker.call(method, url, timeout=self.timeout, data=data, headers={'content-type': 'application/json', 'Api-Key': self.api_key})
+        except pybreaker.CircuitBreakerError as e:
+            logging.getLogger(__name__).error('Geovelo routing service dead (error: {})'.format(e))
+            self.record_external_failure('circuit breaker open')
+        except requests.Timeout as t:
+            logging.getLogger(__name__).error('Geovelo routing service dead (error: {})'.format(t))
+            self.record_external_failure('timeout')
+        except Exception as e:
+            logging.getLogger(__name__).exception('Geovelo routing error')
+            self.record_external_failure(str(e))
+        return None
+
+    def _get_matrix(self, json_response):
+        sn_routing_matrix = response_pb2.StreetNetworkRoutingMatrix()
+        row = sn_routing_matrix.rows.add()
+        for e in json_response:
+            if e[2] == 'duration':
+                continue
+            routing = row.routing_response.add()
+            if e[2]:
+                routing.duration = e[2]
+                routing.routing_status = response_pb2.reached
+            else:
+                routing.duration = -1
+                routing.routing_status = response_pb2.unknown
+        return sn_routing_matrix
+
+    def _check_response(self, response):
+        if response == None:
+            raise TechnicalError('impossible to access geovelo service')
+        if response.status_code != 200:
+            logging.getLogger(__name__).error('Geovelo service unavailable, impossible to query : {}'
```

Comment:
copied on valhalla too... depends on how we handle errors, for now it's fine for me